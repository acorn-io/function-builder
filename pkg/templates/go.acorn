let baseImage: "golang:1.21"
let runtimeImage: "cgr.dev/chainguard/wolfi-base"

args: {
	buildPackages: [string] || default ["."]
}

localFiles: {
	".dockerignore": """
tmp/
"""
	"go.mod": """
module local

go 1.21.5
"""
}

detect: [{
	files: ["go.mod"]
	priority: 100
}, {
	files: ["*.go"]
	priority: 50
}]

targets: dev: {
	from: baseImage
	cacheVolumes: [
		"/root/.cache/go-build",
		"/go/pkg/mod",
	]
	acornfile: """
		dirs: {
			"/root/.cache/go-build": "volume://go-cache-\\(serviceName)?subPath=go-build"
			"/go/pkg/mod": "volume://go-cache-\\(serviceName)?subPath=mod"
			"/src": "./\\(build.context)"
		}
		build: watchFiles: [
			"\\(build.context)/go.mod",
			"\\(build.context)/go.sum",
			"\\(build.context)/vendor/modules.txt",
		]
	"""

	entrypoint: ["go", "run", "github.com/ibuildthecloud/conair@d6bf8ba"]
	run: [
		{copy: dest: "/src"},
	]
}

targets: build: {
	from: baseImage
	cacheVolumes: [
		"/root/.cache/go-build",
		"/go/pkg/mod",
	]
	run: [
		{copy: dest: "/src"},
		"""
        set -x -e
        if [ ! -d vendor ]; then
            go mod tidy
        fi
        go build -o /out/main \(std.join(args.buildPackages, " "))
        """,
	]
}

targets: default: {
	from:       runtimeImage
	workdir:    "/tmp"
	entrypoint: "main"
	run: [
		{
			copy: {
				from:   "build"
				source: "/out/main"
				dest:   "/bin/"
			}
		},
	]
}
