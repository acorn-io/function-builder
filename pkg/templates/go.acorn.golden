{
  "targets": {
    "build": {
      "from": "golang:1.21",
      "cacheVolumes": [
        "/root/.cache/go-build",
        "/go/pkg/mod"
      ],
      "workdir": "/src",
      "run": [
        {
          "copy": {
            "source": ".",
            "dest": "/src"
          }
        },
        {
          "command": "set -x -e\nif [ ! -d vendor ]; then\n    go mod tidy\nfi\ngo build -o /out/main ."
        }
      ]
    },
    "default": {
      "from": "cgr.dev/chainguard/wolfi-base",
      "entrypoint": [
        "main"
      ],
      "workdir": "/tmp",
      "user": 10000,
      "run": [
        {
          "copy": {
            "source": "/out/main",
            "dest": "/bin/",
            "from": "build"
          }
        }
      ]
    },
    "dev": {
      "from": "golang:1.21",
      "cacheVolumes": [
        "/root/.cache/go-build",
        "/go/pkg/mod"
      ],
      "entrypoint": [
        "go",
        "run",
        "github.com/ibuildthecloud/conair@d6bf8ba"
      ],
      "workdir": "/src",
      "acornfile": "\tdirs: {\n\t\t\"/root/.cache/go-build\": \"volume://go-cache-\\(serviceName)?subPath=go-build\"\n\t\t\"/go/pkg/mod\": \"volume://go-cache-\\(serviceName)?subPath=mod\"\n\t\t\"/src\": \"./\\(build.context)\"\n\t}\n\tbuild: watchFiles: [\n\t\t\"\\(build.context)/go.mod\",\n\t\t\"\\(build.context)/go.sum\",\n\t\t\"\\(build.context)/vendor/modules.txt\",\n\t]",
      "run": [
        {
          "copy": {
            "source": ".",
            "dest": "/src"
          }
        }
      ]
    }
  },
  "localFiles": {
    ".dockerignore": "tmp/",
    "go.mod": "module local\n\ngo 1.21.5"
  },
  "detect": [
    {
      "priority": 100,
      "files": [
        "go.mod"
      ]
    },
    {
      "priority": 50,
      "files": [
        "*.go"
      ]
    }
  ]
}